File: lc0146_lruCache/main.py
import typing
from collections import heapq
from collections import OrderedDict

class LRUCacheOld:

    def __init__(self, capacity: int):
        self.cache = dict()
        self.cnt = 0
        self.rused = []
        self.capacity = capacity

    def get(self, key: int) -> int:
        if key in self.cache:
            self.cnt += 1
            self.cache[key] = (self.cache[key][0], self.cnt)
            heapq.heappush(self.rused, (self.cnt, key))
        
        return self.cache.get(key, (-1, -1))[0]

    def put(self, key: int, value: int) -> None:
        if key not in self.cache and len(self.cache) == self.capacity:
            to_evict = -1
            while to_evict == -1:
                to_evict = heapq.heappop(self.rused)
                if to_evict[0] < self.cache[to_evict[1]][1]:
                    to_evict = -1

            del self.cache[to_evict[1]]
        
        self.cnt += 1
        self.cache[key] = (value, self.cnt)
        heapq.heappush(self.rused, (self.cnt, key))


# Your LRUCache object will be instantiated and called as such:

class LRUCache:

    def __init__(self, capacity: int):
        self.cache = OrderedDict()
        self.capacity = capacity

    def get(self, key: int) -> int:
        if key in self.cache:
            self.cache.move_to_end(key)
        
        return self.cache.get(key, -1)

    def put(self, key: int, value: int) -> None:
        while key not in self.cache and len(self.cache) >= self.capacity:
            self.cache.popitem(last=False)
        
        self.cache[key] = value
        self.cache.move_to_end(key)


if __name__ == "__main__":
    sol = LRUCache(20)

    # obj = LRUCache(capacity)
    # param_1 = obj.get(key)
    # obj.put(key,value)

    tests = [
        # Test Cases
    ]

    for t, test in enumerate(tests):
        print(f"======= Test {t} ========")
        answer = sol.lruCache(*test)
        print(answer)

File: lc0146_lruCache/main.cpp
#include <iostream>
#include <string>
#include <vector>
#include <queue>
#include <map>
#include "utils.h"

using namespace std;


class LRUCache {
private: 
    int capacity;
    priority_queue<pair<int, int>, std::vector<pair<int, int> >, std::greater<pair<int, int> >> rused;
    int cnt = 0;
    map<int, pair<int, int> > cache;

public:

    LRUCache(int capacity_) {
        capacity = capacity_;
    }
    
    int get(int key) {
        if (cache.count(key)) {
            cnt++;
            cache[key] = make_pair(cache[key].first, cnt);
            rused.push(make_pair(cnt, key));

            cout << "Get returns:" << cache[key].first << endl;
            cout << "Assigning: " << cnt << " to cache together with value" << endl;
            return cache[key].first;
        }

        return -1;
    }
    
    void put(int key, int value) {
        if (cache.count(key) == 0 && cache.size() == capacity) {
            pair<int, int> to_evict = make_pair(-1, -1);

            // find the first non-stale value on the heap
            while (to_evict.first == -1) {
                to_evict = rused.top();
                rused.pop();
                if (to_evict.first < cache[to_evict.second].second) {
                    to_evict = make_pair(-1, -1);
                }
            }
            
            cout << "Put evicting " << to_evict.first << " " << to_evict.second << endl;
            cache.erase(to_evict.second);
        }
        cnt++;
        cache[key] = make_pair(value, cnt);
        rused.push(make_pair(cnt, key));
        cout << "Put assigns:" << value << " to " << key << endl;
        cout << "Assigning: " << cnt << " to cache together with value" << endl;
    }
};

int main() {
/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */

    int capacity = 2;
    LRUCache* obj = new LRUCache(capacity);

    obj->put(1, 1);
    obj->put(2, 2);
    obj->get(1);
    obj->put(3, 3);
    obj->get(2);
    obj->put(4, 4);
    obj->get(1);
    obj->get(3);
    obj->get(4);

    // int param_1 = obj->get(key);
    // obj->put(key,value);

    return 0;
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */